DataGrid: customize cell/row appearance (#4501)
DevExpress/devextreme-documentation: concepts/05 UI Components/DataGrid/15 Columns/40 Customize Cells/2 Customize Cell Appearance.md

> To customize cell appearance, use the [columns](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/).[cellTemplate](/api-reference/_hidden/dxDataGridColumn/cellTemplate.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#cellTemplate') property or the [onCellPrepared](/api-reference/10%20UI%20Components/dxDataGrid/1%20Configuration/onCellPrepared.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/#onCellPrepared') function. The [columns](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/).[cellTemplate](/api-reference/_hidden/dxDataGridColumn/cellTemplate.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#cellTemplate') option allows you to replace the default cell layout with a custom markup. The [onCellPrepared](/api-reference/10%20UI%20Components/dxDataGrid/1%20Configuration/onCellPrepared.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/#onCellPrepared') event handler allows you to customize the existing layout after it is generated.

Reviewer: I'm just going to comment on the order of things (structure). 

The `cssClass` property seems to be the most basic way to customize cell appearance. If I simply want to change font settings in all grid cells, that is what I would use? At least that is what I would definitely use if I need one column to stand out. No complicated code required, no conditions to be checked. I simply assign a class and that's it. So maybe this simple option should go first? 

The current introduction then mentions `cellTemplate` fist. `onCellPrepared` goes after it. But the rest of the topic reverts this order - you first give an example for `onCellPrepared`, only then for `cellTemplate`. 

Part of the difference between these two method are described in the introduction. Another part of the explanation is next to the second screenshot. Maybe it makes sense to have it all in one place. 

Then we can add some structure to the topic (section titles), like this: 

----------------------

This topic describes how you can customize cell appearance.

##  Apply a Style to All Cells in a Column 

... talk about `cssClass` ...

## Apply a Style to Individual Cells or Rows (Conditional Formatting)

### Change the Markup for the Entire Cell

Handle the onCellPrepared event... It allows you to customize the entire cell's markup after the layout has been generated... 

<example and screenshot>

### Specify a Cell Content Template

Specify a cellTemplate property. It allows you to replace cell content with custom markup, but does not give you access to the entire cell element.

<example and screenshot>

> You can also use CSS rules and assign a class to the [cssClass](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#cssClass) property to customize cell appearance. 

Reviewer: I would probably add something that indicates that it's a column property. 

```suggestion
You can also use CSS rules and assign a class to the [columns.cssClass](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#cssClass) property to customize cell appearance. 
```

                        e.cellElement.classList.add("my-class");
                    }
                }
            }
        }
    }
    </script>

##### React

    <!-- tab: App.js -->
    import React from 'react';

    import 'devextreme/dist/css/dx.light.css';

    import DataGrid from 'devextreme-react/data-grid';

    const cellPrepared = (e) => {
        if (e.rowType === "data") {
            if (e.column.dataField === "Velocity" && e.data.Velocity > e.data.SpeedLimit) {
                e.cellElement.style.cssText = "color: white; background-color: red";
                // or
                e.cellElement.classList.add("my-class");
            }
        }
    }

    function App() {
        return (
            <DataGrid ...
                onCellPrepared={cellPrepared}>
            </DataGrid>
        );
    }

    export default App;
    
---

The example below shows how to use the [columns](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/).[cellTemplate](/api-reference/_hidden/dxDataGridColumn/cellTemplate.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#cellTemplate') property to implement the same task.

> ![Change Color Conditionally with cellTemplate](/images/DataGrid/customize-appearance/column-cellTemplate.png)

Reviewer: The image is not exactly the same as the previous. Maybe this difference requires a comment from us.

### Apply a Style to All Cells in a Column

> You can use CSS rules and assign a class to the [columns](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/).[cssClass](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#cssClass) property to apply styles to all cells in a column. 

Reviewer 2: ```suggestion
You can use CSS rules and assign a class to the [columns](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/).[cssClass](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#cssClass) property to apply styles to all the cells in a column. 
```

### Apply a Style to Individual Cells or Rows (Conditional Formatting)

> Implement the [onCellPrepared](/api-reference/10%20UI%20Components/dxDataGrid/1%20Configuration/onCellPrepared.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/#onCellPrepared') function. It allows you to customize the entire cell's markup after the layout has been generated.

Reviewer 2: ```suggestion
Implement the [onCellPrepared](/api-reference/10%20UI%20Components/dxDataGrid/1%20Configuration/onCellPrepared.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/#onCellPrepared') function. It allows you to customize the entire cell's markup after the layout is generated.
```

> The following example demonstrates how to use the [onCellPrepared](/api-reference/10%20UI%20Components/dxDataGrid/1%20Configuration/onCellPrepared.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/#onCellPrepared') function to change cell color conditionally. As a result, the DataGrid component paints the cells where speed is higher than speed limit red.

Reviewer 2: ```suggestion
The following example demonstrates how to use the [onCellPrepared](/api-reference/10%20UI%20Components/dxDataGrid/1%20Configuration/onCellPrepared.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/#onCellPrepared') function to conditionally change the cell color. As a result, the DataGrid component paints the cells red where the speed exceeds the speed limit.
```

> Remove the `e.column.dataField === "Speed"` condition to apply the appearance to all cells in a row.

Reviewer 2: ```suggestion
Remove the `e.column.dataField === "Speed"` condition to apply the appearance to all the cells in a row.
```

                    }
                }
            }
        }
    }
    </script>

##### React

    <!-- tab: App.js -->
    import React from 'react';

    import 'devextreme/dist/css/dx.light.css';

    import DataGrid from 'devextreme-react/data-grid';

    const cellPrepared = (e) => {
        if (e.rowType === "data") {
            if (e.column.dataField === "Speed" && e.data.Speed > e.data.SpeedLimit) {
                e.cellElement.style.cssText = "color: white; background-color: red";
                // or
                e.cellElement.classList.add("my-class");
            }
        }
    }

    function App() {
        return (
            <DataGrid ...
                onCellPrepared={cellPrepared}>
            </DataGrid>
        );
    }

    export default App;
    
---

You can also specify the [columns](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/).[cellTemplate](/api-reference/_hidden/dxDataGridColumn/cellTemplate.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#cellTemplate') property. It allows you to replace cell content with custom markup, but does not give you access to the entire cell element.

> The example below shows how to use the [columns](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/).[cellTemplate](/api-reference/_hidden/dxDataGridColumn/cellTemplate.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#cellTemplate') property to implement the same task. 

Reviewer 2: `to implement the same task` I'm not sure you can implement a task - a function or feature or functionality, yes

    import 'devextreme/dist/css/dx.light.css';

    import DxDataGrid, {
        DxColumn
    } from 'devextreme-vue/data-grid';

    export default {
        components: {
            DxDataGrid,
            DxColumn
        },
        // ...
    }
    </script>

    <style>
        .my-class {
            background-color: red;
            color: white;
        }
    </style>

##### React

    <!-- tab: App.js -->
    import React from 'react';

    import 'devextreme/dist/css/dx.light.css';

    import DataGrid, {
        Column
    } from 'devextreme-react/data-grid';

    const renderGridCell = (data) => {
        let className = '';
        if (cellInfo.data.Speed > cellInfo.data.SpeedLimit) {
            className += 'my-class';
        }
        return <div className={className}>{cellInfo.value}</div>;
    }

    function App() {
        return (
            <DataGrid ... >
                <Column
                    dataField="Speed"
                    cellRender={renderGridCell}
                />
            </DataGrid>
        );
    }
    export default App;

    <!-- tab: index.css -->
    .my-class {
        background-color: red;
        color: white;
    }

---

> If a customization is not immediately applied, use the [repaint()](/Documentation/ApiReference/UI_Components/dxDataGrid/Methods/#repaint) method to repaint DataGrid or the [repaintRows(rowIndexes)](/Documentation/ApiReference/UI_Components/dxDataGrid/Methods/#repaintRowsrowIndexes) method to repaint specific rows only.

Reviewer 2: ```suggestion
If the changes are not applied immediately, use the [repaint()](/Documentation/ApiReference/UI_Components/dxDataGrid/Methods/#repaint) method to repaint DataGrid or the [repaintRows(rowIndexes)](/Documentation/ApiReference/UI_Components/dxDataGrid/Methods/#repaintRowsrowIndexes) method to repaint specific rows only.
```
